Source/ARTRealtimePresence.m

```diff
@@ -780,85 +778,110 @@ - (void)reenterInternalMembers {
             case ARTPresenceEnter:
             case ARTPresenceUpdate:
-                messageCopy.action = ARTPresencePresent;
-                // intentional fallthrough
             case ARTPresencePresent:
-                [self internalAdd:messageCopy];
+                messageCopy.action = ARTPresencePresent;
+                [self addInternalMember:messageCopy];
                 break;
             case ARTPresenceLeave:
-                [self internalRemove:messageCopy];
+                if (!message.isSynthesized) {
+                    [self removeInternalMember:messageCopy];
+                }
                 break;
             default:
                 break;
         }
-        return YES;
     }
-    ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
-    latest.syncSessionId = _syncSessionId;
-    return NO;
-}
+    
+    BOOL memberUpdated = false;
+    switch (message.action) {
+        case ARTPresenceEnter:
+        case ARTPresenceUpdate:
+        case ARTPresencePresent:
+            messageCopy.action = ARTPresencePresent;
+            memberUpdated = [self addMember:messageCopy];
+            break;
+        case ARTPresenceLeave:
+            if (self.syncInProgress) {
+                messageCopy.action = ARTPresenceAbsent; // RTP2f
+                memberUpdated = [self addMember:messageCopy];
+            } else {
+                memberUpdated = [self removeMember:messageCopy];
+            }
+            break;
+        default:
+            break;
+    }
 
-- (void)internalAdd:(ARTPresenceMessage *)message {
-    [self internalAdd:message withSessionId:_syncSessionId];
+    if (memberUpdated) {
+        [self broadcast:message];
+    }
+    else {
+        ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
+    }
+    [_beforeSyncMembers removeObjectForKey:message.memberKey]; // RTP19
 }
 
-- (void)internalAdd:(ARTPresenceMessage *)message withSessionId:(NSUInteger)sessionId {
-    message.syncSessionId = sessionId;
-    [_members setObject:message forKey:message.memberKey];
-    // Local member
-    if ([message.connectionId isEqualToString:self.connectionId]) {
-        _internalMembers[message.clientId] = message;
-        ARTLogDebug(_logger, @"local member %@ with action %@ has been added", message.memberKey, ARTPresenceActionToStr(message.action).uppercaseString);
+- (BOOL)addMember:(ARTPresenceMessage *)message {
+    ARTPresenceMessage *existing = [_members objectForKey:message.memberKey];
+    if (existing && [existing isNewerThan:message]) {
+        return false;
     }
+    _members[message.memberKey] = message;
+    return true;
 }
 
-- (void)internalRemove:(ARTPresenceMessage *)message {
-    [self internalRemove:message force:false];
+- (BOOL)addInternalMember:(ARTPresenceMessage *)message {
```

Comment:
No need to return values here, since they are not being used right?

Reply:
https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_presence.go#L321

Reply:
https://github.com/ably/ably-cocoa/pull/1891/commits/7ae70b3a8efd93ae93643657d7de0c2abdf4a2a1